src/libopensc/pkcs15-cert.c

```diff
@@ -141,6 +133,195 @@ parse_x509_cert(sc_context_t *ctx, struct sc_pkcs15_der *der, struct sc_pkcs15_c
+		if (*name == NULL) {
+			*name = malloc(dummy_len);
+			if (*name == NULL) {
+				LOG_FUNC_RETURN(ctx, SC_ERROR_OUT_OF_MEMORY);
+			}
+			*name_len = dummy_len;
+		}
+		*name_len = MIN(dummy_len, *name_len);
+		memcpy(*name, dummy, *name_len);
+		LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+	}
+	LOG_FUNC_RETURN(ctx, SC_ERROR_ASN1_OBJECT_NOT_FOUND);
+}
+
+/* get a specific extension from the cert. The extension is identified by it's oid value. NOTE:
+ * extensions can occur in any number or any order, which is why we can't parse them with a single
+ * pass of the asn1 decoder. If is_critical is supplied, then it is set to 1 if the extention is critical and 0
+ * if it is not. The data in the extension is extension specific. The following are
+ * common extension values:
+ *   Subject Key ID: type[] = {0x55, 0x1d, 0x0e} type_len = 3;
+ *   Key Usage: type[] = {0x55, 0x1d, 0x0f} type_len = 3;
+ *   Subject Alt Name: type[] = {0x55, 0x1d, 0x11} type_len = 3;
+ *   Basic Constraints: type[] = {0x55, 0x1d, 0x13} type_len = 3;
+ *   CRL Distribution Points: type[] = {0x55, 0x1d, 0x1f} type_len = 3;
+ *   Certificate Policies: type[] = {0x55, 0x1d, 0x20} type_len = 3;
+ *   Extended Key Usage: type[] = {0x55, 0x1d, 0x25} type_len = 3;
+ * if *name is NULL, sc_pkcs15_get_extension will allocate space for name.
+ */
+int
+sc_pkcs15_get_extension(struct sc_context *ctx, struct sc_pkcs15_cert *cert,
+	const struct sc_object_id *type, u8 **ext_val,
+	size_t *ext_val_len, int *is_critical)
+{
+	const u8 *ext = NULL;
+	const u8 *next_ext = NULL;
+	size_t ext_len = 0;
+	size_t next_ext_len = 0;
+	struct sc_object_id oid;
+	u8 *val;
+	size_t val_len;
+	int critical;
+	int r;
+	struct sc_asn1_entry asn1_cert_ext[] = {
+		{ "x509v3 entry OID", SC_ASN1_OBJECT, SC_ASN1_TAG_OBJECT, 0, &oid, 0 },
+		{ "criticalFlag",  SC_ASN1_BOOLEAN, SC_ASN1_TAG_BOOLEAN, SC_ASN1_OPTIONAL, &critical, NULL },
+		{ "extensionValue",SC_ASN1_OCTET_STRING, SC_ASN1_TAG_OCTET_STRING, SC_ASN1_ALLOC, &val, &val_len },
+		{ NULL, 0, 0, 0, NULL, NULL }
+	};
+
+	for (next_ext = cert->extensions, next_ext_len = cert->extensions_len;
+		next_ext_len; ) {
+
+		/* unwrap the set and point to the next ava */
+		ext = sc_asn1_skip_tag(ctx, &next_ext, &next_ext_len,
+			SC_ASN1_TAG_SEQUENCE | SC_ASN1_CONS, &ext_len);
+		if (ext == NULL) {
+			LOG_TEST_RET(ctx, SC_ERROR_INVALID_ASN1_OBJECT, "ASN.1 decoding of AVA");
+		}
+
+		/*
+		 * use the sc_asn1_decoder for clarity. NOTE it would be more efficient to do this by hand
+		 * so we avoid the man malloc/frees here, but one hopes that one day the asn1_decode will allow
+		 * a 'static pointer' flag that returns a const pointer to the actual asn1 space so we only need
+		 * to make a final copy of the extension value before we return */
+		critical = 0;
+		r = sc_asn1_decode(ctx, asn1_cert_ext, ext, ext_len, NULL, NULL);
+		if (r < 0) {
+			LOG_FUNC_RETURN(ctx, r);
+		}
+
+
+		/* is it the RN we are looking for */
+		if(sc_compare_oid(&oid, type) == 0) {
```

Comment:
Thanks. You are right. I read the docs too fast and the wording in comments confused me:

>  * @return 1 if the oids are equal and a non-zero value otherwise

It should probably say

>  * @return 1 if the oids are equal and a zero value otherwise

Will fix in the next commit